config.go

```diff
@@ -137,10 +137,12 @@ func (c *Config) Interface(num, alt int) (*Interface, error) {
 	}
 
 	// Select an alternate setting if needed (device has multiple alternate settings).
-	if len(c.Desc.Interfaces[num].AltSettings) > 1 {
-		if err := c.dev.ctx.libusb.setAlt(c.dev.handle, uint8(num), uint8(alt)); err != nil {
-			c.dev.ctx.libusb.release(c.dev.handle, uint8(num))
-			return nil, fmt.Errorf("failed to set alternate config %d on interface %d of %s: %v", alt, num, c, err)
+	for i := range c.Desc.Interfaces {
```

Comment:
the `intfDesc` function called in line 123 also iterates over interfaces for the same purpose. See my comment in https://github.com/google/gousb/pull/129#discussion_r1679875622 - instead of looping twice, split `intfDesc` which currently looks up the interface and looks up an alternate setting descriptor (i.e. it does two things) into two distinct helpers - one that looks up the interface, another that looks up the alt descriptor. Splitting into two helpers will give you a variable here that will hold the correct interface descriptor, it won't be necessary to look up the interface again.

Reply:
I am pushing back against this idea as that method doesn't seem to be providing all of the variables needed. I think this is as far as we can go with using altInfo:

```go
for i := range c.Desc.Interfaces {
	if altInfo.Number == num && len(c.Desc.Interfaces[i].AltSettings) > 1 {
		if err := c.dev.ctx.libusb.setAlt(c.dev.handle, uint8(num), uint8(alt)); err != nil {
			c.dev.ctx.libusb.release(c.dev.handle, uint8(num))
			return nil, fmt.Errorf("failed to set alternate config %d on interface %d of %s: %v", alt, num, c, err)
		}
	}
}
```

Suggestions welcome.

Reply:
Perhaps I'm not communicating myself clearly. But perhaps the simplest option would be - if you provide me with the output of `lsusb -v` for the affected device, I'm happy to make the necessary changes and add tests for that use case.

Reply:
So the output from `lsusb` is as follows:

```
Bus 002 Device 001: ID 1278:0525 1278 ECHO2 
Bus 000 Device 000: ID 1278:0525 1278 USB 3.1 Bus 
Bus 000 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub 
```

and

```
USB:

    USB 3.1 Bus:

      Host Controller Driver: AppleT8122USBXHCI

    USB 3.1 Bus:

      Host Controller Driver: AppleT8122USBXHCI

        ECHO2:

          Product ID: 0x0525
          Vendor ID: 0x1278
          Version: 0.00
          Speed: Up to 480 Mb/s
          Manufacturer: Starlight Xpress
          Location ID: 0x02100000 / 1
          Current Available (mA): 500
          Current Required (mA): 300
          Extra Operating Current (mA): 0

    USB 3.1 Bus:

      Host Controller Driver: AppleT8122USBXHCI
```

Let me know if you need anything else ...

Reply:
`lsusb -v` please (with the descriptor data dump)

Reply:
The above is the output.

Reply:
ah, I see, lsusb on MacOS apparently is not the same `lsusb` from `usbutils` as on Linux. I'm not sure how to get the descriptor info dump natively on MacOS - a quick search suggests no good answer. 

In that case, please run the `lsusb` that comes with `gousb` (`github.com/google/gousb/lsusb`). I hope this one at least still works and provides useful info...

Reply:
I'm travelling today, so without device. But I will get this over to you ASAP (probably this evening, 21:00BST).

Reply:
any updates?

Reply:
@zagrodzki Strange, I only looked at this today and realised I haven’t sent it back to you. I will do it this weekend.

Reply:
ping again. With another pull request in, I'll be publishing a new release in ~2wks, would be great if this fix was included.

Reply:
@zagrodzki On this now ... apologies for the delay. I've been on holiday.

Reply:
@zagrodzki Hopefully this is useful:

```
002.001 1278:0525 Unknown (Starlight Xpress)
  Protocol: (Defined at Interface level)
  Configuration 1:
    --------------
    Interface 1 alternate setting 0 (available endpoints: [0x01(1,OUT) 0x82(2,IN)])
      Vendor Specific Class
      ep #2 IN (address 0x82) bulk [512 bytes]
      ep #1 OUT (address 0x01) bulk [512 bytes]
    --------------
 ```

Reply:
thanks, that confirms my suspicion of the root cause:

```
Interface 1 alternate setting 0 (available endpoints: [0x01(1,OUT) 0x82(2,IN)])
```

means the bInterfaceNumber is set to 1, but it should be 0 according to the specs. The code tries to access interface descriptors index 1, which doesn't exist.